golang/cosmos/proto/agoric/swingset/swingset.proto

```diff
@@ -167,3 +189,66 @@ message SwingStoreArtifact {
 
   bytes data = 2 [(gogoproto.jsontag) = "data", (gogoproto.moretags) = "yaml:\"data\""];
 }
+
+// ChunkedArtifact is the manifest for an artifact that is submitted across
+// multiple transactions, in chunks, as when using InstallBundle to submit
+// chunks.
+message ChunkedArtifact {
+  // The SHA-512 hash of the entire bundle file's contents.
+  string sha512 = 1 [(gogoproto.jsontag) = "sha512", (gogoproto.moretags) = "yaml:\"sha512\""];
+
+  // The size of the final bundle artifact in bytes.
+  uint64 size_bytes = 2 [(gogoproto.jsontag) = "size_bytes", (gogoproto.moretags) = "yaml:\"size_bytes\""];
+
+  // Information about the chunks that will be concatenated to form this
+  // bundle.
+  repeated ChunkInfo chunks = 3 [(gogoproto.jsontag) = "chunks", (gogoproto.moretags) = "yaml:\"chunks\""];
+}
+
+// Current state of this chunk.
+enum ChunkState {
+  // Unknown state.
+  CHUNK_STATE_UNSPECIFIED = 0;
+
+  // The chunk is still in-flight.
+  CHUNK_STATE_IN_FLIGHT = 1;
+
+  // The chunk has been received.
+  CHUNK_STATE_RECEIVED = 2;
+
+  // The chunk has been processed.
+  CHUNK_STATE_PROCESSED = 3;
+};
+
+// Information about a chunk of a bundle.
+message ChunkInfo {
+  // The SHA-512 hash of the chunk contents.
+  string sha512 = 1 [(gogoproto.jsontag) = "sha512", (gogoproto.moretags) = "yaml:\"sha512\""];
+
+  // The chunk size in bytes.
+  uint64 size_bytes = 2 [(gogoproto.jsontag) = "size_bytes", (gogoproto.moretags) = "yaml:\"size_bytes\""];
+
+  // The current state of the chunk.
+  ChunkState state = 3 [(gogoproto.jsontag) = "state", (gogoproto.moretags) = "yaml:\"state\""];
+}
+
+// A node in a doubly-linked-list of chunks of a chunked artifact, as used for
+// chunked bundle installation, in order of ascending block time.
+// The keeper uses this to expediently expire stale chunks.
+message ChunkedArtifactNode {
+  // The id of the pending bundle installation.
+  uint64 chunked_artifact_id = 1
+      [(gogoproto.jsontag) = "chunkedArtifactId", (gogoproto.moretags) = "yaml:\"chunkedArtifactId\""];
+
+  // Doubly-linked list.
+  uint64 next_id = 2 [(gogoproto.jsontag) = "nextId", (gogoproto.moretags) = "yaml:\"nextId\""];
+
+  uint64 prev_id = 3 [(gogoproto.jsontag) = "prevId", (gogoproto.moretags) = "yaml:\"prevId\""];
```

Comment:
```suggestion
  // The ID of the next chunked artifact in the list.
  uint64 next_id = 2 [(gogoproto.jsontag) = "nextId", (gogoproto.moretags) = "yaml:\"nextId\""];

  // The ID of the previous chunked artifact in the list.
  uint64 prev_id = 3 [(gogoproto.jsontag) = "prevId", (gogoproto.moretags) = "yaml:\"prevId\""];
```
The doc comments should also clarify edge case behavior (e.g., `chunked_artifact_id` and `next_id` and `prev_id` might all have the same value for the sole node of a list, or alternatively list start/end could be indicated by empty `prev_id`/`next_id` [respectively]).